brave/src/main/java/brave/propagation/ExtraField.java

```diff
@@ -0,0 +1,330 @@
+ * spanCustomizer.tag(countryCode.name(), countryCode.getValue());
+ * spanCustomizer.tag(countryCode.name(), countryCode.getValue(context));
+ * spanCustomizer.tag("country-code", ExtraField.getValue("country-code"));
+ *
+ * // You can also set or override the value similarly, which might be needed if a new request
+ * countryCode.setValue("FO");
+ * countryCode.setValue(context, "FO");
+ * ExtraField.setValue("country-code", "FO");
+ * ExtraField.setValue(context, "country-code", "FO");
+ * }</pre>
+ *
+ * <h3>Correlation</h3>
+ * If you want an extra field to also be available in correlation such as logging contexts, use
+ * {@link Builder#withCorrelation()}.
+ *
+ * <pre>{@code
+ * // configure the field, permitting it to be used in correlation contexts
+ * amznTraceId = ExtraField.newBuilder("x-amzn-trace-id").withCorrelation().build();
+ *
+ * // Allow logging patterns like %X{traceId} %X{x-amzn-trace-id}
+ * loggingContext = new Log4J2Context();
+ *
+ * decorator = CorrelationFieldScopeDecorator.newBuilder(loggingContext)
+ *                                           .addField(amznTraceId).build();
+ *
+ * tracingBuilder.propagationFactory(ExtraFieldPropagation.newFactoryBuilder(B3Propagation.FACTORY)
+ *                                                        .addField(amznTraceId)
+ *                                                        .build())
+ *               .currentTraceContext(ThreadLocalCurrentTraceContext.newBuilder()
+ *                                                                  .addScopeDecorator(decorator)
+ *                                                                  .build())
+ * }</pre>
+ *
+ * <h3>Prefixed fields</h3>
+ * <p>You can also prefix fields, if they follow a common pattern. For example, the following will
+ * propagate the field "x-vcap-request-id" as-is, but send the fields "country-code" and "user-id"
+ * on the wire as "baggage-country-code" and "baggage-user-id" respectively.
+ *
+ * <pre>{@code
+ * requestId = ExtraField.create("x-vcap-request-id");
+ * userId = ExtraField.newBuilder("user-id").prefix("baggage-").build();
+ * countryCode = ExtraField.newBuilder("country-code").prefix("baggage-").build();
+ *
+ * // Later, you can call below to affect the country code of the current trace context
+ * ExtraField.setValue("country-code", "FO");
+ * String countryCode = ExtraField.getValue("country-code");
+ * }</pre>
+ *
+ * <h3>Appropriate usage</h3>
+ * It is generally not a good idea to use the tracing system for application logic or critical code
+ * such as security context propagation.
+ *
+ * <p>Brave is an infrastructure library: you will create lock-in if you expose its apis into
+ * business code. Prefer exposing your own types for utility functions that use this class as this
+ * will insulate you from lock-in.
+ *
+ * <p>While it may seem convenient, do not use this for security context propagation as it was not
```

Comment:
if the paragraph about the infrastructure library gets removed consider merging this one and the `it is generally not a good idea ...` into one.